cpp/src/search/contains_nested.cu

```diff
@@ -59,4 +65,87 @@ bool contains_nested_element(column_view const& haystack,
+      rmm::exec_policy(stream), out_begin, out_begin + needles.size(), false);
+    return result;
+  }
+
+  auto const haystack_tv        = table_view{{haystack}};
+  auto const needles_tv         = table_view{{needles}};
+  auto const haystack_has_nulls = has_nested_nulls(haystack_tv);
+  auto const needles_has_nulls  = has_nested_nulls(needles_tv);
+
+  auto haystack_map =
+    detail::hash_map_type{compute_hash_table_size(haystack.size()),
+                          detail::COMPACTION_EMPTY_KEY_SENTINEL,
+                          detail::COMPACTION_EMPTY_VALUE_SENTINEL,
+                          detail::hash_table_allocator_type{default_allocator<char>{}, stream},
+                          stream.value()};
+
+  // Insert all indices of the elements in the haystack column into the hash map.
+  // As such, we will use `thrust::equal_to` as key comparator to not ignore any key.
+  //
+  // An alternative way to this is to use `self_comparator` for key comparisons, which would only
+  // insert unique rows of the haystack column. This would save some memory (or significant amount
+  // of memory if the haystack column contains many duplicate rows), however, will result in much
+  // more processing time due to expensive row comparisons.
+  {
+    auto const haystack_it = cudf::detail::make_counting_transform_iterator(
+      size_type{0}, [] __device__(size_type const i) { return cuco::make_pair(i, i); });
+
+    auto const hasher   = cudf::experimental::row::hash::row_hasher(haystack_tv, stream);
+    auto const d_hasher = detail::experimental::compaction_hash(
+      hasher.device_hasher(nullate::DYNAMIC{haystack_has_nulls}));
+
+    haystack_map.insert(haystack_it,
+                        haystack_it + haystack.size(),
+                        d_hasher,
+                        thrust::equal_to<size_type>{},
+                        stream.value());
+  }
+
+  // Check for existence of needles in haystack.
+  // During this, we will use `negative_index_comparator_adapter` to convert the existing indices
+  // in the hash map (i.e., indices of haystack elements) into `lhs_index_type`, and indices of the
+  // searching needles into `rhs_index_type` for row comparisons.
+  {
+    // A reverse iterator constructed from `0` value will begin from `-1`.
+    // Thus, needle indices will iterate in reverse order in the range `[-1, -1-needles.size())`.
+    // They will be converted back to the range `[0, needles.size())` then into `rhs_index_type`
+    // automatically by `negative_index_hasher_adapter` and `negative_index_comparator_adapter`.
+    auto const needles_it =
+      thrust::make_reverse_iterator(thrust::make_counting_iterator(size_type{0}));
+
+    auto const hasher   = cudf::experimental::row::hash::row_hasher(needles_tv, stream);
+    auto const d_hasher = cudf::experimental::row::hash::negative_index_hasher_adapter{
+      detail::experimental::compaction_hash(
+        hasher.device_hasher(nullate::DYNAMIC{needles_has_nulls}))};
```

Comment:
(After I spent a bunch of time describing this implementation, I realized that I had just described the implementation of `semi_join`. This whole `contains` algorithm should just be a trivial call to [`semi_join`](https://github.com/rapidsai/cudf/blob/447028f25c9c00e26c1a8e1d3765de7a565ee58c/cpp/include/cudf/join.hpp#L318-L349))

This negative index stuff is far more complicated than what is needed here. 

You can probe with a type that is distinct from the `key_type` of the map. I explained this in the last standup about "heterogeneous lookup".  

Furthermore, you can do a lot better here then simply inserting the row indices. Instead, you use the hash value of the row as the key and the index of the row as the payload. 

```
cuco::static_multimap<hash_value_type, rhs_index_t> map{...}; // need to use multimap because map doesn't have pair_contains (yet)

auto const haystack_hasher = row_hasher(haystack, stream);
auto haystack_hashes = make_counting_transform_iterator(0, haystack_hasher.device_hasher());
auto haystack_hash_and_index = thrust::zip_iterator{haystack_hashes, rhs_iterator{0}};

map.insert(haystack_hash_and_index, haystack_hash_and_index + haystack.size(), {}, {}, stream);

auto const needle_hasher = row_hasher(needles, stream);
auto needle_hashes = make_counting_transform_iterator(0, needle_hasher.device_hasher());
auto needle_hash_and_index = thrust::zip_iterator{needle_hashes, lhs_iterator{0}};

auto comparator = strong_index_comparator{...};

map.pair_contains(needle_hash_and_index, needle_hash_and_index + needles.size(), out_begin, {}, 
[comparator] __device__ (auto lhs_hash_and_index, auto rhs_hash_and_index){
   auto const& [lhs_hash, lhs_index] = lhs_hash_and_index;
   auto const& [rhs_hash, rhs_index] = rhs_hash_and_index;
   if(lhs_hash == rhs_hash){
       return comparator(lhs_index, rhs_index);
   }
   return false;
});
```

Reply:
Alright, so this can share implementation with `semi_join`. However, it may be a bit out of scope of this PR. Instead, it should rather be some issue like "support nested types in semi_join". Addressing it will refactor this out to share the implementation. Does that sound good @jrhemstad?

Reply:
> You can probe with a type that is distinct from the key_type of the map. I explained this in the last standup about "heterogeneous lookup".

I have implemented the negative index solution before our meeting. After that, I tried "heterogeneous lookup" but maybe I missed some key components thus I couldn't get it works anyhow. So I just continued using my negative idx solution.

Reply:
I believe that we need multimap because we want to insert all hash values. But why if we have `pair_contains` then we will deprecate it soon?

In addition, I'm not sure if the approach @jrhemstad suggested above can overcome the output type restriction: We have to use the same payload value for both inserting and containing check. Again, "heterogeneous lookup" is needed.

Reply:
I only used a multimap because it's the only one that has the `pair_*` functions. I didn't realize it didn't have a `pair_contains` yet :) 

More info on [heterogeneous lookup](https://www.cppstories.com/2021/heterogeneous-access-cpp20/). cuCo's doesn't work exactly the same, but the main idea of probing with a different type than the `value_type` of the container is the same.

Reply:
Humm, the idea of such lookup relies on having a lookup function with template key type, as I expected:
```
// two functions:
bool contains( const Key& key ) const;

template <class K> 
bool contains() const K& x ) const; // enabled when Hash::is_transparent and
                                    // KeyEqual::is_transparent is present
```

But I see in cuco `static_map`, `static_map::contains` internally calls `device_view::contains` which has just 2 versions but require a fixed type for key:
```
__device__ bool contains(Key const& k,...)
```

So I'm not sure how to apply it here without modifying the internal implementation of `static_map`. Although `static_map::contains` doxygen says that:
```
* @tparam InputIt Device accessible input iterator whose `value_type` is convertible to the map's `key_type`
```
but internally `value_type` is not converted *explicitly* to `key_type`. Instead, `key_type` is used directly:
```
writeBuffer[threadIdx.x] = view.contains(key, hash, key_equal);
```

If we have something like this then I think we are good:
```
writeBuffer[threadIdx.x] = view.contains(static_cast<Key>(key), hash, key_equal);
```

Reply:
@PointKernel Do you think this is doable? (`static_map_kernels.cuh`, line 457).

Reply:
Ah, you're right @ttnghia. We'll need to modify the `contains` function to allow the queried key to be a template type. 

@PointKernel will work on that.

Reply:
Thanks. I have filed the related issues: 
 * https://github.com/NVIDIA/cuCollections/issues/172
 * https://github.com/NVIDIA/cuCollections/issues/173

Reply:
@ttnghia Can you please confirm the priority between `static_multimap::pair_contains`, template `static_map::contains` and template `static_multimap::contains`?

Reply:
If we **can** have `pair_contains`, we must have template key types for `static_multimap::pair_contains`. I need both in my final implementation.

But I can get unblocked if we just have template key types for `static_map::contains` or `static_multimap::contains` implemented. Then I will switch to use `pair_contains` later when it is ready.

Reply:
@jrhemstad Can you explain more why using `pair_contains` is better here, please?

Reply:
By using the "row hash value + row index" logic, we compute the row hash once (for the insertion) and do one row comparison only if hash values are equal during retrieval (`pair_contains`). If we use the single map and insert row index only, we need to compute the expensive row hash twice (one for insertion and another for retrieval) and perform the expensive row comparison all the time during retrieval.